lib/FrontendTool/FrontendTool.cpp

```diff
@@ -337,6 +521,13 @@ static void computeSwiftModuleTraceInfo(
                              // this is good enough.
         : buffer.str();
 
+      /// FIXME: We need the name-check hack here because abiImportedModules
+      /// (sometimes?) ends up only having the Clang module and no Swift module
+      /// for overlays when the overlay is not explicitly imported in Swift.
```

Comment:
Ah, I didn't notice that it had an optional parameter. Good point. I will try to use that.

---

Somewhat of a tangent but it looks like the optional parameter is only used in two places overall, both of which are in the cross-import overlay code. 😅 I'm wondering if we should remove the optional parameter and split it into two-three methods: `getTopLevelModule`, `getClangOverlayOrTopLevelModule` and maybe a `getClangOverlay` (if needed)?

With that, most places would continue calling `getTopLevelModule`.

The following code would call `getClangOverlayOrTopLevelModule`.

```
    // If we are importing a submodule, treat it as though we imported its
    // top-level module (or rather, the top-level module's clang overlay if it
    // has one).
    if (isSubmodule(nextImport.importedModule)) {
      nextImport.importedModule =
          nextImport.importedModule->getTopLevelModule(/*overlay=*/true);
```

This code looks like it should have an error if it ends up not getting an overlay, is that right? If yes, it could call `getClangOverlay`. If not, it could call `getClangOverlayOrTopLevelModule`.

```
  if (I.getUnderlyingModule()) {
    auto underlying = I.getUnderlyingModule().get();

    // If this is a clang module, and it has a clang overlay, we want the
    // separately-imported overlay to sit on top of the clang overlay.
    if (underlying->isNonSwiftModule())
      underlying = underlying->getTopLevelModule(true);

    // FIXME: Should we warn if M doesn't reexport underlyingModule?
    SF.addSeparatelyImportedOverlay(M, underlying);
  }
```

Having the name + optional-parameter seems a bit confusing from an API perspective. It does kinda' make sense in a way, because the Clang overlay sits "on top of" the underlying module, but I feel like having a more verbose name would clarify the intent better.

Wdyt? [If you agree, I will file a starter bug since this seems like a doable refactoring for a newcomer.]

Reply:
The second snippet would need to use `getClangOverlayOrTopLevelModule()`; it's meant to be a no-op if there's no clang overlay.

The code is structured using a flag because it allowed the most code to be shared between the two operations. However, I'm not necessarily averse to splitting `getTopLevelModule()` in two—it just means a little bit of duplication.

Reply:
Using the `overlay` parameter in the right place does avoid the name check hack. Neat!